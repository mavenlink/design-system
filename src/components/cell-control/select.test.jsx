import React, { createRef } from 'react';
import { render as _render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import Select from './select.jsx';
import ListOption from '../list-option/list-option.jsx';

const baseListOptions = ['foo', 'bar'];
const baseListOptionRefs = baseListOptions.map(() => createRef());
const baseListOptionElements = ({ onSelect }) => baseListOptions.map((option, index) => (
  <ListOption key={option} onSelect={onSelect} ref={baseListOptionRefs[index]} value={option}>
    {option}
  </ListOption>
));

const render = (ui, options = { labelledBy: 'labelled-by' }) => (
  _render(ui, {
    ...options,
    wrapper: props => (
      <table>
        <thead>
          <tr>
            <th id={options.labelledBy}>Column Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            {props.children}
          </tr>
        </tbody>
      </table>
    ),
  })
);

describe('Select cell control', () => {
  const requiredProps = {
    id: 'test-id',
    labelledBy: 'labelled-by',
    listOptionRefs: [],
    name: 'test-name',
  };

  it('has defaults', () => {
    const ref = createRef();
    render(<Select {...requiredProps} ref={ref} />);
    expect(document.body).toMatchSnapshot();
    expect(document.body).toBe(document.activeElement);
    expect(ref.current).toMatchSnapshot();
  });

  it('className API', () => {
    render(<Select {...requiredProps} className="unique-class-name" />);
    expect(document.body).toMatchSnapshot();
  });

  it('accepts a value', () => {
    const value = 'foo';
    render(<Select {...requiredProps} value={value}>{baseListOptionElements}</Select>);
    expect(screen.getByRole('combobox')).toHaveValue('foo');
  });

  describe('onChange API', () => {
    it('calls onChange when a new value is selected', () => {
      const onChange = jest.fn();

      render(<Select {...requiredProps} onChange={onChange}>{baseListOptionElements}</Select>);

      user.click(screen.getByRole('combobox'));
      user.click(screen.getByText('foo'));

      expect(onChange).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({ value: 'foo' }),
      }));
    });
  });

  describe('validationMessage API', () => {
    it('shows a tooltip on the invalid icon', () => {
      render(<Select {...requiredProps} validationMessage="This is an error message." />);
      expect(screen.getByRole('combobox')).toHaveAccessibleDescription('This is an error message.');
      user.hover(screen.getByRole('img', { name: 'This is an error message.' }));
      expect(document.body).toMatchSnapshot();
    });
  });
});
