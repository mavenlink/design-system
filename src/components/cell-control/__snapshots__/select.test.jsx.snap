// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select cell control className API 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="unique-class-name"
            role="gridcell"
          >
            <div
              class="container"
              style="position: relative;"
            >
              <input
                aria-autocomplete="none"
                aria-controls="test-id-single-choice-listbox"
                aria-describedby="test-id-tooltip test-idHint"
                aria-expanded="false"
                aria-haspopup="listbox"
                autocomplete="off"
                class="input"
                id="test-id"
                name="test-name"
                role="combobox"
                style="--numIcon: 3;"
                type="text"
                value=""
              />
              <div
                class="container"
              >
                <svg
                  aria-disabled="false"
                  class="button"
                  height="6"
                  role="button"
                  tabindex="0"
                  width="10"
                >
                  <title>
                    Open choices listbox
                  </title>
                  <use
                    xlink:href="#caret-down.svg"
                  />
                </svg>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Select cell control has defaults 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="container"
            role="gridcell"
          >
            <div
              class="container"
              style="position: relative;"
            >
              <input
                aria-autocomplete="none"
                aria-controls="test-id-single-choice-listbox"
                aria-describedby="test-id-tooltip test-idHint"
                aria-expanded="false"
                aria-haspopup="listbox"
                autocomplete="off"
                class="input"
                id="test-id"
                name="test-name"
                role="combobox"
                style="--numIcon: 3;"
                type="text"
                value=""
              />
              <div
                class="container"
              >
                <svg
                  aria-disabled="false"
                  class="button"
                  height="6"
                  role="button"
                  tabindex="0"
                  width="10"
                >
                  <title>
                    Open choices listbox
                  </title>
                  <use
                    xlink:href="#caret-down.svg"
                  />
                </svg>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Select cell control has defaults 2`] = `
Object {
  "dirty": false,
  "name": "test-name",
  "value": undefined,
}
`;

exports[`Select cell control validationMessage API shows a tooltip on the invalid icon 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="container"
            role="gridcell"
          >
            <div
              class="container"
              style="position: relative;"
            >
              <input
                aria-autocomplete="none"
                aria-controls="test-id-single-choice-listbox"
                aria-describedby="test-id-tooltip test-idHint"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-invalid="true"
                autocomplete="off"
                class="invalidInput"
                id="test-id"
                name="test-name"
                role="combobox"
                style="--numIcon: 3;"
                type="text"
                value=""
              />
              <div
                class="container"
              >
                <div
                  class="wrapper"
                >
                  <svg
                    height="18"
                    role="img"
                    width="22"
                  >
                    <title
                      id="test-idHint"
                    >
                      This is an error message.
                    </title>
                    <use
                      xlink:href="#caution.svg"
                    />
                  </svg>
                </div>
                <svg
                  aria-disabled="false"
                  class="button"
                  height="6"
                  role="button"
                  tabindex="0"
                  width="10"
                >
                  <title>
                    Open choices listbox
                  </title>
                  <use
                    xlink:href="#caret-down.svg"
                  />
                </svg>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <span
    class="tooltip left"
    id=""
    style="--triangle-height: 8px; left: -128px; top: 0px;"
  >
    This is an error message.
  </span>
</body>
`;
